### PR TITLE
Refactor SmartSim API docs

### PR DESCRIPTION
**Pull request by amandarichardsonn**
This PR addresses the refactor of the SmartSim API documentation on craylabs.org

Currently up for review
1. Experiment.rst file -> Entities section, Launcher section

### CODE REVIEW

File: doc/experiment.rst [hunk 1]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
 
-  settings = exp.create_run_settings("echo", exe_args="Hello World")
-  model = exp.create_model("hello_world", settings)
+SmartSim currently supports 5 `launchers`:
+  1. ``local``: for single-node, workstation, or laptop
+  2. ``slurm``: for systems using the Slurm scheduler
+  3. ``pbs``: for systems using the PBSpro scheduler
+  4. ``cobalt``: for systems using the Cobalt scheduler
+  5. ``lsf``: for systems using the LSF scheduler
+  6. ``auto``: have SmartSim auto-detect the launcher to use.
```
Comment (al-rigazzi) [inline doc/experiment.rst:63]: Standardize punctuation on list items - I have no preference but would like to see consistency across list items and lists

File: doc/experiment.rst [hunk 2]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
 
-  exp.start(model, block=True)
-  print(exp.get_status(model))
+If `launcher="auto"` is used, the experiment will attempt to find a launcher
+on the system, and use the first one it encounters. If a launcher cannot
```
Comment (al-rigazzi) [inline doc/experiment.rst:66]: Maybe "encounters" -> "detects"?

File: doc/experiment.rst [hunk 3]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+be found or no launcher parameter is provided, the default value of
+`launcher="local"` will be used.
 
-If the launcher has been specified, or auto-detected through setting
-``launcher=auto`` in the Experiment initialization, the ``create_run_settings``
-method will automatically create the appropriate ``RunSettings`` object and
-return it.
+A `launcher` enables SmartSim to interact with the compute system
+programmatically through the Experiment python user interface.
```
Comment (al-rigazzi) [inline doc/experiment.rst:71]: Capitalize "Python"

File: doc/experiment.rst [hunk 4]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+SmartSim users do not have to leave the Jupyter Notebook,
```
Comment (al-rigazzi) [inline doc/experiment.rst:72]: I understand what you mean, but I think it would be easier to follow it if it was slightly rephrased - something similar to (feel free to modify, this is just a suggestion)
"Thanks to the ``Experiment`` API, users can launch, query, and interact with their jobs from a single Jupyter Notebook, Python REPL, or Python script."

File: doc/experiment.rst [hunk 5]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+Python REPL, or Python script to launch, query, and interact with their jobs.
 
-For example with Slurm
+To specify a specific launcher, one argument needs to be provided
```
Comment (al-rigazzi) [inline doc/experiment.rst:75]: Maybe this sentence (and the code block) could be moved up - I think it could go before the preceding sentence, and right after the explanation of default behavior.

File: doc/experiment.rst [hunk 6]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
-
-The above will run ``srun -n 32 -N 1 echo Hello World!``, monitor its
-execution, and inform the user when it is completed. This driver script can be
-executed in an interactive allocation, or placed into a batch script as follows:
+.. code-block:: python
 
-.. code-block:: bash
+    from smartsim import Experiment
+
+    exp = Experiment("name-of-experiment", launcher="local")  # local launcher
```
Comment (al-rigazzi) [inline doc/experiment.rst:82]: I think we may want to just show one or two examples, and possibly in two separated code blocks, introduced by a sentence, like: "for example, to set up a local launcher, the ``Experiment`` would be specified as follows." and then maybe just again the explanation of ``auto`` or of a WLM. I think you explained thoroughly that ``launcher`` is the parameter one has to set.

File: doc/experiment.rst [hunk 7]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+    exp = Experiment("name-of-experiment", launcher="slurm")  # Slurm launcher
+    exp = Experiment("name-of-experiment", launcher="pbs")    # PBSpro launcher
+    exp = Experiment("name-of-experiment", launcher="cobalt") # Cobalt launcher
+    exp = Experiment("name-of-experiment", launcher="lsf")    # LSF launcher
+    exp = Experiment("name-of-experiment", launcher="auto")   # auto-detect launcher
+
+
+=========
+ Entities
```
Comment (al-rigazzi) [inline doc/experiment.rst:91]: There is a space before `Entities`. Unless it is intentional, I'd delete it (just in case some interpreter complains about indentation).

File: doc/experiment.rst [hunk 8]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+``RunSettings``, and ``BatchSettings``, along with a discussion of their relationships.
+
+Entity Relationship Diagram
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Below is an Entity Relationship Diagram using Crow's Foot Notation.
+Moving from the top to the bottom of the diagram, you will first notice that
+an Experiment can have none or many ``Orchestrators``, ``Models``, and ``Ensembles``.
+The Experiment class is responsible for all entity initialization functions:
+
+1. ``Experiment.create_orchestrator()`` :raw-html:`&rarr; ``Orchestrator``
```
Comment (al-rigazzi) [inline doc/experiment.rst:108]: The `raw-html` directive is not your best option here. I would suggest to add `..    include:: <isonum.txt>` at the top of the file, and then use `|rarr|` instead of ``:raw-html:`&rarr`` (which also misses the `` ` `` at the end)

File: doc/experiment.rst [hunk 9]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
 
-    #!/bin/bash
-    #SBATCH --exclusive
-    #SBATCH --nodes=1
-    #SBATCH --ntasks-per-node=32
-    #SBATCH --time=00:10:00
+|SmartSim ERD|
+
+Moving on to the ``Model`` entity class, you will notice that a ``Model`` requires a
+minimum and maximum of one ``RunSetting`` object. However, it accepts a minimum of zero
```
Comment (al-rigazzi) [inline doc/experiment.rst:121]: I think "a minimum and maximum of one" can be replaced by "exactly one", while "a minimum of zero and a maximum of one" can be replaced by "at most one", to keep the tone more colloquial.

File: doc/experiment.rst [hunk 10]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+and a maximum of one ``BatchSetting`` object.
+
+The ``Ensemble`` entity class does not require either ``RunSetting`` or ``BatchSetting``
+objects upon initialization but does accept a maximum of one for each. INCORRECT
```
Comment (al-rigazzi) [inline doc/experiment.rst:125]: OK, I see "INCORRECT", thus I assume you're going to change this sentence.

File: doc/experiment.rst [hunk 11]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+
+Orchestrator
+^^^^^^^^^^^^
+The ``Orchestrator`` is an in-memory database that can be launched alongside
+``Model`` and ``Ensemble`` entities in SmartSim. The ``Orchestrator`` can be used to store and retrieve
```
Comment (al-rigazzi) [inline doc/experiment.rst:130]: If you're going to refer to the orchestrator as a "concept", I think you can write it with lowercase O in plain text. If you refer to the class, then ``Orchestrator`` is correct.

File: doc/experiment.rst [hunk 12]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+data during the course of an experiment. In order to stream data into or
+receive data from the Orchestrator, one of the SmartSim clients
+(SmartRedis) has to be used within a Model.
+
+To create a standard orchestrator, use the ``Experiment.create_database()``
+function. A standard orchestrator will run the database and applications on
+separate nodes. The database can be single or multi-sharded.
+For a colocated orchestrator, use the ``model.colocated_db()``
```
Comment (al-rigazzi) [inline doc/experiment.rst:138]: I think it may be easier to follow this paragraph if you first introduced the concept of colocated vs clustered DB and then explained how to spin up one or the other. Also, I think you can just refer to the colocated one as database and not orchestrato (and we will then update the rest accordingly).

File: doc/experiment.rst [hunk 13]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+function provided by the Model object. This action results in
```
Comment (al-rigazzi) [inline doc/experiment.rst:139]: Model -> ``Model``

File: doc/experiment.rst [hunk 14]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+function mandates the specification of a unique database identifier, denoted
+by the `db_identifier` argument, per created orchestrator.
 
-    python /path/to/script.py
+Model
+^^^^^
+Models represent any computational kernel, including applications,
+scripts, or generally, a program. They are flexible enough
+to support various applications; however, to be used with our
+clients (SmartRedis), the application must be written in Python,
```
Comment (al-rigazzi) [inline doc/experiment.rst:154]: I am not sure that we should specify this here, but if we do, then maybe I'd rephrase it to make it clear that if one wants an application launched as `Model` to use SmartRedis clients, then the application must be written in Python, C, C++, or Fortran. Also, I would not refer to SmartRedis as to "our clients", but simply as SmartRedis.

File: doc/experiment.rst [hunk 15]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+C, C++, or Fortran.
+
+A Model is created through the function ``Experiment.create_model()``.
+During initialization, models are given RunSettings objects that specify
+how a kernel should be executed with regard to the workload manager
+(e.g., Slurm) and the available compute resources on the system.
+Optionally, the user may also specify a BatchSettings object if
+the model should be launched as a batch on the WLM system.
```
Comment (al-rigazzi) [inline doc/experiment.rst:162]: batch -> batch job

File: doc/experiment.rst [hunk 16]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
 
 Ensemble
-========
-
-In addition to a single model, SmartSim has the ability to launch an
+^^^^^^^^
+In addition to a single model, SmartSim has the ability to launch a
```
Comment (al-rigazzi) [inline doc/experiment.rst:166]: a -> an?

**Review by al-rigazzi — COMMENTED** [review]
Left some comments, I will add some more. In general I like the logical flow, we may just need to add more code snippets.

File: doc/experiment.rst [hunk 17]
```diff
@@ -12,316 +38,211 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+receive data from the Orchestrator, one of the SmartSim clients
+(SmartRedis) has to be used within a Model.
+
+To create a standard orchestrator, use the ``Experiment.create_database()``
+function. A standard orchestrator will run the database and applications on
+separate nodes. The database can be single or multi-sharded.
+For a colocated orchestrator, use the ``model.colocated_db()``
+function provided by the Model object. This action results in
+the launch of an orchestrator on the same node as the application
+but is limited to single-sharded database configuration.
```
Comment (mellis13) [inline doc/experiment.rst:141]: I think having nice graphics for the different deployment options for the Orchestrator would be helpful for users.  Let's adapt some existing graphics we might have.

File: doc/experiment.rst [hunk 18]
```diff
@@ -12,316 +38,211 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+
+SmartSim supports multi-database functionality, enabling an experiment
+to have several concurrently launched database instances. If there is a
```
Comment (mellis13) [inline doc/experiment.rst:144]: database instances-> Orchestrators

File: doc/experiment.rst [hunk 19]
```diff
@@ -12,316 +38,211 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+need to launch more than one ``Orchestrator``, the ``Experiment.create_database()``
+function mandates the specification of a unique database identifier, denoted
+by the `db_identifier` argument, per created orchestrator.
```
Comment (mellis13) [inline doc/experiment.rst:147]: Maybe clarify at the end of this paragraph:

"The `db-identifier` is used to make it easier to reference SmartSim Orchestrator(s) from application client code.  This is particularly useful in instances where an Orchestrator is colocated with a ``SmartSim`` model for low-latency inference and another Orchestrator is launched to handle other aspects of the workflow such as visualization and ML model training.  More detailed information on the ideal use cases for Orchestrators and co-located Orchestrators are available in sections <link to Orchestrator sections>"

File: doc/experiment.rst [hunk 20]
```diff
@@ -12,316 +38,211 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+``Ensemble`` will create the underlying model objects. An ensemble is created
+with the function ``Experiment.create_ensemble()`` and requires
+one of the subsequent sets of arguments upon initialization:
+1. run_settings and params
+2. run_settings and replicas
+3. batch_settings
+4. batch_settings, run_settings, and params
+5. batch_settings, run_settings, and replicas
+
+RunSettings
```
Comment (mellis13) [inline doc/experiment.rst:178]: As discussed, I don't think RunSettings and BatchSettings should be in this section.

File: doc/experiment.rst [hunk 21]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
-function should take two arguments: a list of parameter names, and a list of
-lists of potential parameter values. The function should return a list of
-dictionaries that will be supplied as model parameters. The length of the list
-returned will determine how many ``Model`` instances are created.
-
-For example, the following is the built-in strategy ``all_perm``:
+``Ensemble`` will create the underlying model objects. An ensemble is created
+with the function ``Experiment.create_ensemble()`` and requires
+one of the subsequent sets of arguments upon initialization:
+1. run_settings and params
```
Comment (al-rigazzi) [inline doc/experiment.rst:172]: run_settings -> `run_settings`
params -> `params`

And so on for the rest of the list

File: doc/experiment.rst [hunk 22]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+entity during stage initialization via the `run_settings` parameter.
+When creating a ``RunSettings`` object
+via the ``Experiment.create_run_settings()`` function, the appropriate ``RunSettings``
+object will be return based on what WLM you initialized the experiment with.
+See here to view the list of offered RunSettings WLM objects.
+
+BatchSettings
+^^^^^^^^^^^^^
+``BatchSettings`` is used to configure entities that should be launched
+as a batch on a WLM system. The ``BatchSettings`` object is applied to an
```
Comment (al-rigazzi) [inline doc/experiment.rst:194]: batch -> batch job

File: doc/experiment.rst [hunk 23]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+entity during stage initialization via the `batch_settings` parameter.
```
Comment (al-rigazzi) [inline doc/experiment.rst:195]: Not sure I follow what "stage initialization" is - was it introduced anywhere in the docs?

File: doc/experiment.rst [hunk 24]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+When creating a ``BatchSettings`` object
+via the ``Experiment.create_batch_settings()`` function, the appropriate ``BatchSettings``
+object will be return based on what WLM you initialized the experiment with.
```
Comment (al-rigazzi) [inline doc/experiment.rst:198]: return -> returned

File: doc/experiment.rst [hunk 25]
```diff
@@ -12,316 +38,231 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+See here to view the list of offered RunSettings WLM objects.
```
Comment (al-rigazzi) [inline doc/experiment.rst:199]: RunSettings -> `RunSettings`

And then please add the link (to the API, I think)

File: doc/experiment.rst [hunk 26]
```diff
@@ -12,316 +38,211 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+SmartSim supports multi-database functionality, enabling an experiment
+to have several concurrently launched database instances. If there is a
+need to launch more than one ``Orchestrator``, the ``Experiment.create_database()``
+function mandates the specification of a unique database identifier, denoted
+by the `db_identifier` argument, per created orchestrator.
 
-    python /path/to/script.py
+Model
+^^^^^
+Models represent any computational kernel, including applications,
```
Comment (mellis13) [inline doc/experiment.rst:151]: Maybe it would be more positive as:

"Models represent any computational kernel, including applications, scripts, or generally, a program. They can interact with other SmartSim entities via data transmitted to/from SmartSim Orchestrators.  Clients for connecting to a Orchestrator are available in Python, C++, C, and Fortran."

**Review by al-rigazzi — COMMENTED** [review]
Added a few last comments on the sections you restructured

File: doc/experiment.rst [hunk 27]
```diff
@@ -12,316 +38,211 @@ Jupyter notebook, interactive python session, or Python file and use the
...
+how a kernel should be executed with regard to the workload manager
+(e.g., Slurm) and the available compute resources on the system.
+Optionally, the user may also specify a BatchSettings object if
+the model should be launched as a batch on the WLM system.
 
 Ensemble
-========
-
-In addition to a single model, SmartSim has the ability to launch an
+^^^^^^^^
```
Comment (mellis13) [inline doc/experiment.rst:165]: This section doesn't seem to capture the essence and purpose of Ensembles.  I think we we need to add language around how it allows users to launch parameterized groups of Models and through that we explain other items like runsettings, params, etc.

**Review by mellis13 — CHANGES_REQUESTED** [review]
Review in progress, but here is some initial feedback

File: doc/experiment.rst (hunk 16 above)
Comment (amandarichardsonn) [reply doc/experiment.rst:166]: I think 'a' is correct?
